feat: Added Project object to Feast Objects

EXPEbdodla opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. Ensure that your code follows our code conventions: https://github.com/feast-dev/feast/blob/master/CONTRIBUTING.md#code-style-and-linting
2. Run unit tests and ensure that they are passing: https://github.com/feast-dev/feast/blob/master/CONTRIBUTING.md#unit-tests
3. If your change introduces any API changes, make sure to update the integration tests here: https://github.com/feast-dev/feast/tree/master/sdk/python/tests
4. Make sure documentation is updated for your PR!
5. Make sure your commits are signed: https://github.com/feast-dev/feast/blob/master/CONTRIBUTING.md#signing-off-commits
6. Make sure your PR title follows conventional commits (e.g. fix: [Description of ...], feat: [Description of ...], chore: [Description of ...], refactor: [Description of ...])

-->

# What this PR does / why we need it:
1. Based on PR #4425 discussion, Added support for Project object for all registry's in feature definitions. Defaults to project in feature_store.yaml if not specified. Syncs projects content in registry (table) from feast_metadata on initialization. Opens up an option to specify multiple projects in single repository (Needs more work). 
2. Added a registry configuration `purge_feast_metadata` to stop using `feast_metadata` table because list_project_metadata may be used by clients. Instead of deleting that, added a configuration to stop populating table and switch to projects specific methods. 
3. Refactored registry methods `_prepare_registry_for_changes` and `_get_registry_proto` methods in registry.py. Refactored proto() methods in sql.py and snowflake.py. 
4. Corrected RegistryConfig model. __new__() method in registry.py is blocking RegistryConfig inheritance. Removed it and corrected the tests accordingly. Now we can defined configs specific to registry instead of defining all at RegistryConfig class. 

# Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
-->


# Misc
<!--
Feel free to leave additional thoughts or tag people as you see fit
-->

tokoko commented on sdk/python/feast/errors.py:
```diff
@@ -477,6 +477,16 @@ def __init__(self, name, project=None):
             super().__init__(f"Permission {name} does not exist")
 
 
+class ProjectNotFoundException(Exception):
```
Why are both of these necessary? what's the difference. btw, All Feast exceptoins should extend `FeastError`.

dmartinol replied:
@tokoko I noticed the same while doing the latest refactory but I didn't want to extend the scope of changes too much. Looks like other duplications are there like `PermissionObjectNotFoundException` and `DataSourceObjectNotFoundException`. Can we remove all of them (in a dedicated ticket maybe)?

EXPEbdodla replied:
I'll extend the FeastError. 
Reasons for having two exceptions: Just followed the current convention. 
ProjectNotFoundException --> Exception thrown when the DELETE calls are made and project is not found. 
ProjectObjectNotFoundException --> Exception thrown when GET calls are made and project is not found. _get_object method passes two args (name, project) when object is not found. I don't want to add another if else logic to filter out project.

tokoko replied:
@dmartinol yeah, this just looks weird. Probably best to defer removal until after this is merged, though.. just to avoid a messy rebase

dmartinol replied:
in case, remember an item in the next Release Notes to remind clients to replace any try/except that may have used those duplicate errors.

tokoko commented:
FYI @dmartinol

EXPEbdodla commented on sdk/python/feast/repo_config.py:
```diff
@@ -129,11 +130,10 @@ class RegistryConfig(FeastBaseModel):
-
-    cache_mode: StrictStr = "sync"
-    """ str: Cache mode type, Possible options are sync and thread(asynchronous caching using threading library)"""
+    purge_feast_metadata: StrictBool = False
```
@tokoko Do you have any recommendation on how we should do these type of changes?  I just don't want to delete project_metadata in one go.

tokoko replied:
from an API standpoint, since we are still below 1.0 release mark, the policy is to mark as deprecated and then remove after 2-3 minor releases.

The problem in this case is that the deletion should happen not just from API but rather there needs to be changes made in the existing registry as well. I think we will eventually need a new command for this, something like `feast migrate` (borrowing the name from airflow) which should be run after every version bump?

EXPEbdodla replied:
I handled the case for changes to existing registry using `_sync_feast_metadata_to_projects_table` method. Something like `migrate` may help rather than embedding the change in code.

EXPEbdodla commented:
Fixed Integration tests. Removed [test_remote_online_store.py](https://github.com/feast-dev/feast/pull/4475#diff-beab92b9d6c966ff5f8911a97b744554e64cd9e0c8d66b1a3122a751e2e46967) apply step because Client Store Registry Apply brings back the server and client registry store states to the contents in the repository (Repository missing Permissions. Permissions are applied through store.apply() method). Seems like bug in previous versions which let it work. Did some refactoring to registry so it's showing up as error now.

EXPEbdodla commented:
@tokoko Please take a look at this PR when you get a chance.

tokoko commented:
looks good, can you also add usage example in [quickstart](https://github.com/feast-dev/feast/blob/master/docs/getting-started/quickstart.md), just an example of creating a project and specifying some tags.

EXPEbdodla commented:
> looks good, can you also add usage example in [quickstart](https://github.com/feast-dev/feast/blob/master/docs/getting-started/quickstart.md), just an example of creating a project and specifying some tags.

Updated quick start document and `feast init` functionality to include Project object.

tokoko reviewed (approved):
lgtm, thanks
